src/encoded/static/components/matrix_mouse_development.js

```diff
@@ -0,0 +1,1165 @@
+                        subcategory: colSubcategoryBucket.key,
+                    };
+                    colIndex += 1;
+                });
+            }
+        }
+    });
+    return colMap;
+};
+
+/**
+ * Takes matrix data from JSON and generates an object that <DataTable> can use to generate the JSX
+ * for the matrix. This is a shim between the experiment or audit data and the data <DataTable>
+ * needs.
+ * @param {object} context Matrix JSON for the page
+ * @param {func}   getRowCategories Returns rowCategory info including desired display order
+ * @param {func}   getRowSubCategories Returns subCategory desired display order
+ * @param {func}   mapRowCategoryQueries Callback to map row category query values
+ * @param {func}   mapSubCategoryQueries Callback to map subcategory query values
+ * @param {array}  expandedRowCategories Names of rowCategories the user has expanded
+ * @param {func}   expanderClickHandler Called when the user expands/collapses a row category
+ *
+ * @return {object} Generated object suitable for passing to <DataTable>
+ */
+
+const convertExperimentToDataTable = (context, getRowCategories, getRowSubCategories, mapRowCategoryQueries, mapSubCategoryQueries, expandedRowCategories, expanderClickHandler, clearClassifications, stageFilter) => {
+    const rowCategory = context.matrix.y.group_by[0];
+    const rowSubcategory = context.matrix.y.group_by[1];
+
+    // Generate the mapping of column categories and subcategories.
+    const colCategory = context.matrix.x.group_by[0];
+    const colSubcategory = context.matrix.x.group_by[1];
+
+    // Generate the mapping of column categories and subcategories.
+    const colMap = generateNewColMap(context);
+    const colCount = Object.keys(colMap).length;
+
+    // Convert column map to an array of column map values sorted by column number for displaying
+    // in the matrix header.
+    const sortedCols = Object.keys(colMap).map(assayColKey => colMap[assayColKey]).sort((colInfoA, colInfoB) => colInfoA.col - colInfoB.col);
+
+    const colCategoryNames = [];
+    sortedCols.forEach((col) => {
+        if (col.subcategory) {
+            colCategoryNames.push(`${col.category}|${col.subcategory}`);
+        } else if (col.hasSubcategories !== true) {
+            colCategoryNames.push(col.category);
+        }
+    });
+
+    // Generate array of names of assays that have targets and don't collapse their targets, for
+    // rendering those columns as disabled.
+    const colCategoriesWithSubcategories = Object.keys(colMap).filter(colCategoryName => colMap[colCategoryName].hasSubcategories && !collapsedAssays.includes(colCategoryName));
+
+    const { rowCategoryData, rowCategoryNames } = getRowCategories();
+    const rowKeys = ['column-categories'];
+
+    // loop through row data to combine gendered mouse stage/ages into single row
+    const combinedRowCategoryData = JSON.parse(JSON.stringify(rowCategoryData));
```

Comment:
I think I really do need to clone rowCategoryData. This code was very tricky actually. The problem is that I'm using the original rowCategoryData in order to generate the new rowCategoryData and if you don't deep clone the original, you end up with infinite rows and the browser crashes, and it's no good. I agree that this generally isn't optimal but couldn't see a much better way to do this. I think the cloning is kind of tricky because the object has a lot of depth. I looked at the function you suggested and am not sure if it will help, I'll look again on Monday.